lightning/src/onion_message/messenger.rs

```diff
@@ -243,6 +244,90 @@ impl OnionMessageRecipient {
 	}
 }
 
+/// The `ReceivedOnionMessage` enum represents messages received by the `OnionMessageHandler`.
+///
+/// This enum encompasses two variants:
+/// - `WithReplyPath`: Represents a message that can be responded to, including the original message,
+///   an optional path id, and a responder.
+/// - `WithoutReplyPath`: Represents a message without the possibility of response because of absence of a reply_path.
+///   It contains the original message and an optional path id.
+///
+pub enum ReceivedOnionMessage<T>
+where
+	T: OnionMessageContents
+{
+	WithReplyPath {
+		message: T,
+		path_id: Option<[u8; 32]>,
+		responder: Responder<T>,
+	},
+	WithoutReplyPath {
+		message: T,
+		path_id: Option<[u8; 32]>,
+	},
+}
+
+impl<T> ReceivedOnionMessage<T>
+where
+	T: OnionMessageContents
+{
+	fn new( message: T, reply_path_opt: Option<BlindedPath>, path_id: Option<[u8; 32]> ) -> Self {
+		if let Some(reply_path) = reply_path_opt {
+			let responder = Responder {
+				reply_path,
+				_phantom: PhantomData
+			};
+			ReceivedOnionMessage::WithReplyPath {
+				message,
+				path_id,
+				responder
+			}
+		} else {
+			ReceivedOnionMessage::WithoutReplyPath {
+				message,
+				path_id
+			}
+		}
+    }
+}
+
+/// The `Responder` struct manages responses to [`ReceivedOnionMessage`]
+pub struct Responder<T>
+where
+	T: OnionMessageContents
+{
+	reply_path: BlindedPath,
+
+	/// A marker field, ensuring compile-time enforcement that the message
+	/// type used in the struct definition matches the response type.
+	_phantom: PhantomData<T>
+}
+
+impl<T> Responder<T>
+where
+	T: OnionMessageContents
+{
+    pub fn respond(self, response: T) -> ResponseInstruction<T>{
+		ResponseInstruction::HaveResponse {
+			response,
+			reply_path: self.reply_path,
+		}
+	}
+}
+
+/// The `ResponseInstruction` enum represents instructions for responding to [`ReceivedOnionMessage`].
+///
+/// It includes the following variants:
+/// - `HaveResponse`: Indicates that a response is available, and can be sent using reply_path.
+/// - `NoResponse`: Indicates that no response is available.
+pub enum ResponseInstruction<T: OnionMessageContents> {
```

Comment:
Yeah, we also don't want to give the handler the option to return a response if the sender didn't supply a reply path. Although, in the current formulation that is still possible because `ResponseInstruction::HaveResponse` could in theory be constructed by the handler with an arbitrary `BlindedPath`. So it would be better to have that variant wrap a struct that only `Responder` can construct. Then that struct can be re-used in a third variant (next PR) for signaling a reply path should be included with the response.

> Would this be simpler as a struct (and the handle_*message methods would return an `Option`)? That way `respond` would just return a struct containing a message, path pair and the user can just pass those to send_message.

Hmm... I think we want some sort of instructions as it can include context for `OnionMessenger` to log. Like the corresponding `path_id` that it is responding to, which come to think of it is why we probably want to include that in the responder so it can be passed along to the `ResponseInstruction`. Right now it is fine without it for the sync case but we'll need it for the async case.